optuna/samplers/_nsga2/crossover.py

```diff
@@ -0,0 +1,505 @@
+    # Section 4.2 Prototype Algorithm for UNDX-m
+
+    def _normal(rng: np.random.RandomState, ds: List[np.ndarray]) -> np.ndarray:
+        # Compute an orthonormal basis by adding one appropriate vector to ds,
+        # and extract one vector from the orthonormal basis.
+
+        d = rng.normal(0, 1, size=ds[0].shape[0])
+        ds.append(d)
+        X = np.stack(ds)
+        Q, _ = np.linalg.qr(X.T)
+        return Q.T[-1]
+
+    def _orthonormal_basis_vector_from_ds(ds: List[np.ndarray]) -> np.ndarray:
+        # Compute an orthonormal basis of the subspace orthogonal to d_i(1,..i,..m).
+
+        X = np.stack(ds)
+        Q, _ = np.linalg.qr(X.T)
+        return Q.T[-1]
+
+    assert xs.ndim == 2
+    # In this case, `len(xs)==4` because `n_select` is fixed at 4.
+
+    x_mp2, xs = xs[-1], xs[:-1]  # Section 4.2 (1), (3).
+    m = len(xs) - 1
+    dim = len(x_mp2)
+
+    p = np.sum(xs, axis=0) / (m + 1)  # Section 4.2 (2).
+    ds = [x - p for x in xs]  # Section 4.2 (2).
+    n = _normal(rng, ds[:-1])  # Normal to the plane that contains d_i(1,..,m) section 4.2 (4).
+    d_mp2 = x_mp2 - p  # Section 4.2 (4).
+    D = np.dot(d_mp2, n) / np.linalg.norm(n)  # Section 4.2 (4).
+    es = _orthonormal_basis_vector_from_ds(
+        ds[:-1]
+    )  # Orthonormal basis of the subspace orthogonal to d_i(1,..,m) section 4.2 (5).
+
+    one = p  # Section 4.2 (6).
+
+    ws = rng.normal(0, sigma_xi ** 2, size=m)
+    two = np.zeros(dim)  # Section 4.2 (6).
+    for i in range(m):
+        two += ws[i] * ds[i]
+
+    three = np.zeros(dim)  # Section 4.2 (6).
+    for i in range(dim - m):
+        vs = rng.normal(0, sigma_eta ** 2, size=dim)
+        three += vs - np.dot(vs, es[i]) * es[i]
+    three *= D
+    child_params_array = one + two + three
+    return child_params_array
+
+
+def _spx(xs: np.ndarray, rng: np.random.RandomState, epsilon: float) -> np.ndarray:
+    # https://www.researchgate.net/publication/2388486_Progress_Toward_Linkage_Learning_in_Real-Coded_GAs_with_Simplex_Crossover
+    # Section 2 A Brief Review of SPX
+
+    assert xs.ndim == 2
+    # In this case, `len(xs)==3` because `n_select` is fixed at 3.
+
+    n = xs.shape[0] - 1
+    G = xs.sum(axis=0) / xs.shape[0]  # Equation (1).
+    rs = [np.power(rng.uniform(0, 1), 1 / (k + 1)) for k in range(n)]  # Equation (2).
+    xks = [G + epsilon * (pk - G) for pk in xs]  # Equation (3).
+    ck = 0  # Equation (4).
+    for k in range(1, n + 1):
+        ck = rs[k - 1] * (xks[k - 1] - xks[k] + ck)
+
+    child_params_array = xks[-1] + ck  # Equation (5).
+    return child_params_array
+
+
+def _is_contained(params: Dict[str, Any], search_space: Dict[str, BaseDistribution]) -> bool:
+    contains = True
+    for param_name in params.keys():
+        param, param_distribution = params[param_name], search_space[param_name]
+
+        if not param_distribution._contains(param_distribution.to_internal_repr(param)):
+            contains = False
+            break
+    return contains
```

Comment:
The following is more simple, maybe.
```suggestion
def _is_contained(params: Dict[str, Any], search_space: Dict[str, BaseDistribution]) -> bool:
    for param_name in params.keys():
        param, param_distribution = params[param_name], search_space[param_name]

        if not param_distribution._contains(param_distribution.to_internal_repr(param)):
            return False
    return True
```